pkg/funky/router.go

```diff
@@ -0,0 +1,199 @@
+///////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+///////////////////////////////////////////////////////////////////////
+package funky
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// FirstPort the starting port number for servers created by a Router
+const FirstPort uint16 = 9000
+
+// Router an interface for delegating function invocations to idle servers
+type Router interface {
+	Delegate(input Message) (*Message, error)
+	Shutdown() error
+}
+
+// DefaultRouter a struct that hold servers that can be delegated to
+type DefaultRouter struct {
+	servers []Server
+	mutex   *sync.Mutex
+	sem     *semaphore.Weighted
+}
+
+// NewRouter constructor for DefaultRouters
+func NewRouter(numServers int, serverFactory ServerFactory) (*DefaultRouter, error) {
+	if numServers < 1 {
+		return nil, IllegalArgumentError("numServers")
+	}
+
+	servers, err := createServers(numServers, serverFactory)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DefaultRouter{
+		servers: servers,
+		mutex:   &sync.Mutex{},
+		sem:     semaphore.NewWeighted(int64(numServers)),
+	}, nil
+}
+
+// Delegate delegates function invocation to an idle server
+func (r *DefaultRouter) Delegate(input Message) (*Message, error) {
+	server, err := r.findFreeServer()
+	if err != nil {
+		return nil, err
+	}
+
+	defer r.releaseServer(server)
```

Comment:
This should probably look like this:
```go
	defer func() {
		r.releaseServer(server)
	}()
```

The intent being to NOT capture the value of `server` at the time of the `defer` statement: this way `server` can refer to a replacement server if we need to kill the original one if it's unresponsive.